x-pack/plugins/endpoint/public/embeddables/resolver/view/index.tsx

```diff
@@ -0,0 +1,246 @@
+      };
+    }, [handleMouseUp]);
+
+    useEffect(() => {
+      window.addEventListener('mousemove', handleMouseMove, { passive: true });
+      return () => {
+        window.removeEventListener('mousemove', handleMouseMove);
+      };
+    }, [handleMouseMove]);
+
+    // TODO, handle mouse up when no longer on element or event window. ty
+
+    const dotPositions = useMemo(
+      (): ReadonlyArray<readonly [number, number]> => [
+        [0, 0],
+        [0, 100],
+        [100, 100],
+        [100, 0],
+        [100, -100],
+        [0, -100],
+        [-100, -100],
+        [-100, 0],
+        [-100, 100],
+      ],
+      []
+    );
+
+    const refCallback = useCallback(
+      (node: null | HTMLDivElement) => {
+        setRef(node);
+        clientRectCallback(node);
+      },
+      [clientRectCallback]
+    );
+
+    useEffect(() => {
+      // Set the 'wheel' event listener directly on the element
+      // React sets event listeners on the window and routes them back via event propagation. As of Chrome 73 or something, 'wheel' events on the 'window' are automatically treated as 'passive'. Seems weird, but whatever
+      if (ref !== null) {
+        ref.addEventListener('wheel', handleWheel);
+        return () => {
+          ref.removeEventListener('wheel', handleWheel);
+        };
+      }
+    }, [handleWheel, ref]);
+
+    return (
+      <div className={className} ref={refCallback} onMouseDown={handleMouseDown}>
+        {dotPositions.map((worldPosition, index) => (
+          <DiagnosticDot key={index} worldPosition={worldPosition} />
+        ))}
+      </div>
+    );
+  })
+)`
+  /* TODO, this is not a concern of Resolver. its parent needs to do this probably? */
+  display: flex;
+  flex-grow: 1;
+  position: relative;
+`;
+
+/**
+ * Returns a DOMRect sometimes, and a `ref` callback. Put the `ref` as the `ref` property of an element, and
+ * DOMRect will be the result of getBoundingClientRect on it.
+ * Updates automatically when the window resizes. TODO: better Englishe here
+ */
+function useAutoUpdatingClientRect(): [DOMRect | undefined, (node: Element | null) => void] {
+  const [rect, setRect] = useState<DOMRect>();
+  const nodeRef = useRef<Element>();
+
+  const ref = useCallback((node: Element | null) => {
+    // why do we have to deal /w both null and undefined? TODO
+    nodeRef.current = node === null ? undefined : node;
+    if (node !== null) {
+      setRect(node.getBoundingClientRect());
+    }
+  }, []);
+
+  // TODO, this isn't really a concern of Resolver.
+  // The parent should inform resolver that it needs to rerender
+  useEffect(() => {
+    window.addEventListener('resize', handler, { passive: true });
+    return () => {
+      window.removeEventListener('resize', handler);
+    };
+    function handler() {
+      if (nodeRef.current !== undefined) {
+        setRect(nodeRef.current.getBoundingClientRect());
+      }
+    }
+  }, []);
+  return [rect, ref];
+}
+
+const DiagnosticDot = styled(
```

Comment:
@bkimmel here's where a component can find its position relative to the parent resolver component